repair/repair.cc

```diff
@@ -1989,26 +2024,81 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
+                        if (alt_dc_for_keyspace.empty()) {
+                            msg = fmt::format("{}: found no alternative datacenter", msg);
+                            if (source_dc.force()) {
+                                rlogger.warn("{}: using source_dc anyway according to the force option", msg);
+                            } else if (source_dc.user_provided()) {
+                                throw std::runtime_error(format("{}: Omit the source_dc option to use all nodes in the cluster, or use the --force option to enforce it", msg));
+                            } else {
+                                rlogger.warn("{}: falling back to repair using all nodes", msg);
+                                source_dc_for_keyspace = "";
+                            }
+                        }
+                    }
+                    break;
+                }
+                case locator::replication_strategy_type::everywhere_topology:
+                    // source_dc_live_nodes is not empty, so we can use any remaining nodes
+                    source_dc_for_keyspace = *source_dc;
+                    break;
+                case locator::replication_strategy_type::simple:
+                    // With simple strategy, we have no assurance that source_dc will contain
+                    // another replica for all token ranges.
+                    break;
+                default:
+                    break;
+                }
+            }
+            rlogger.info("{}: started with keyspace={}, source_dc={}, alt_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc_for_keyspace, alt_dc_for_keyspace, ranges.size() * nr_tables, ignore_nodes);
             for (auto it = ranges.begin(); it != ranges.end();) {
                 auto& r = *it;
                 seastar::thread::maybe_yield();
                 auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
-                auto neighbors = boost::copy_range<std::vector<gms::inet_address>>(strat.calculate_natural_ips(end_token, *tmptr).get() |
-                    boost::adaptors::filtered([myip, &source_dc, &topology, &ignore_nodes] (const gms::inet_address& node) {
+                auto natural_ips = strat.calculate_natural_ips(end_token, *tmptr).get();
+                auto neighbors = boost::copy_range<std::vector<gms::inet_address>>(natural_ips |
+                    boost::adaptors::filtered([&] (const gms::inet_address& node) {
                         if (node == myip) {
                             return false;
                         }
                         if (ignore_nodes.contains(node)) {
                             return false;
                         }
-                        return source_dc.empty() ? true : topology.get_datacenter(node) == source_dc;
+                        return source_dc_for_keyspace.empty() || live_nodes_per_dc[source_dc_for_keyspace].contains(node)
```

Comment:
source_dc_for_keyspace could be non empty, but live_nodes_per_dc[source_dc_for_keyspace] is not initialized because it is only initialized when find_alternative_datacenter is called in the if branch above.